README.md
# A new version supports React 18

# React Calendar Timeline

A modern and responsive React timeline component.

![calendar demo](https://raw.githubusercontent.com/namespace-ee/react-calendar-timeline/master/demo.gif)

Checkout the [examples here](https://github.com/namespace-ee/react-calendar-timeline/tree/master/examples)!

# Contents

- [Getting Started](#getting-started)
- [Usage](#usage)
- [API](#api)
- [Timeline Markers](#timeline-markers)
- [Timeline Headers](#timeline-headers)
- [FAQ](#faq)
- [Contribute](#contribute)

# Getting started

```bash
# via yarn
yarn add react-calendar-timeline-v2

# via npm
npm install --save react-calendar-timeline-v2
```

`react-calendar-timeline` has [react](https://reactjs.org/), [react-dom](https://reactjs.org/docs/react-dom.html), [`moment`](http://momentjs.com/) and [`interactjs`](http://interactjs.io/docs/) as peer dependencies.

# Usage

At the very minimum:

```jsx
import Timeline from 'react-calendar-timeline-v2';
// make sure you include the timeline stylesheet or the timeline will not be styled
import 'react-calendar-timeline-v2/lib/Timeline.css';
import moment from 'moment';

const groups = [
  { id: 1, title: 'group 1' },
  { id: 2, title: 'group 2' },
];

const items = [
  {
    id: 1,
    group: 1,
    title: 'item 1',
    start_time: moment(),
    end_time: moment().add(1, 'hour'),
  },
  {
    id: 2,
    group: 2,
    title: 'item 2',
    start_time: moment().add(-0.5, 'hour'),
    end_time: moment().add(0.5, 'hour'),
  },
  {
    id: 3,
    group: 1,
    title: 'item 3',
    start_time: moment().add(2, 'hour'),
    end_time: moment().add(3, 'hour'),
  },
];

ReactDOM.render(
  <div>
    Rendered by react!
    <Timeline
      groups={groups}
      items={items}
      defaultTimeStart={moment().add(-12, 'hour')}
      defaultTimeEnd={moment().add(12, 'hour')}
    />
  </div>,
  document.getElementById('root'),
);
```

# API

_NB!_ All props need to be immutable. For example, this means if you wish to change the title of one of your items, please pass in a whole new items array instead of changing the title in the old array. [Here's more info.](http://reactkungfu.com/2015/08/pros-and-cons-of-using-immutability-with-react-js/)

The component can take many props:

## groups

Expects either a vanilla JS array or an immutableJS array, consisting of objects with the following attributes:

```js
{
  id: 1,
  title: 'group 1',
  rightTitle: 'title in the right sidebar',
  stackItems?: true,
  height?: 30
}
```

If you use the right sidebar, you can pass optional `rightTitle` property here.
If you want to overwrite the calculated height with a custom height, you can pass a `height` property as an int in pixels here. This can be very useful for categorized groups.

## items

Expects either a vanilla JS array or an immutableJS array, consisting of objects with the following attributes:

```js
{
  id: 1,
  group: 1,
  title: 'Random title',
  start_time: 1457902922261,
  end_time: 1457902922261 + 86400000,
  canMove: true,
  canResize: false,
  canChangeGroup: false,
  itemProps: {
    // these optional attributes are passed to the root <div /> of each item as <div {...itemProps} />
    'data-custom-attribute': 'Random content',
    'aria-hidden': true,
    onDoubleClick: () => { console.log('You clicked double!') },
    className: 'weekend',
    style: {
      background: 'fuchsia'
    }
  }
}
```

The preferred (fastest) option is to give Unix timestamps in milliseconds for `start_time` and `end_time`. Objects that convert to them (JavaScript `Date` or `moment()`) will also work, but will be a lot slower.

## defaultTimeStart and defaultTimeEnd

Unless overridden by `visibleTimeStart` and `visibleTimeEnd`, specify where the calendar begins and where it ends. This parameter expects a Date or moment object.

## visibleTimeStart and visibleTimeEnd

The exact viewport of the calendar. When these are specified, scrolling in the calendar must be orchestrated by the `onTimeChange` function. This parameter expects a Unix timestamp in milliseconds.

**Note that you need to provide either `defaultTimeStart/End` or `visibleTimeStart/End` for the timeline to function**

## buffer

a number (default to 3) which represents the extra timeline rendered on right and lift of the visible area which the user will scroll through before the time rerenders.

more explication in section [behind the scenes](#behind-the-scenes)

Note: setting buffer to 1 will disable the scrolling on the timeline

## selected

An array with id's corresponding to id's in items (`item.id`). If this prop is set you have to manage the selected items yourself within the `onItemSelect` handler to update the property with new id's and use `onItemDeselect` handler to clear selection. This overwrites the default behaviour of selecting one item on click.

## keys

An array specifying keys in the `items` and `groups` objects. Defaults to

```js
{
  groupIdKey: 'id',
  groupTitleKey: 'title',
  groupRightTitleKey: 'rightTitle',
  itemIdKey: 'id',
  itemTitleKey: 'title',    // key for item div content
  itemDivTitleKey: 'title', // key for item div title (<div title="text"/>)
  itemGroupKey: 'group',
  itemTimeStartKey: 'start_time',
  itemTimeEndKey: 'end_time',
}
```

## className

Additional class names as a string for the root Timeline element.

## sidebarWidth

Width of the sidebar in pixels. If set to `0`, the sidebar is not rendered. Defaults to `150`.

## sidebarContent

Everything passed here will be displayed above the left sidebar. Use this to display small filters or so. Defaults to `null`.

## rightSidebarWidth

Width of the right sidebar in pixels. If set to `0`, the right sidebar is not rendered. Defaults to `0`.

## rightSidebarContent

Everything passed here will be displayed above the right sidebar. Use this to display small filters or so. Defaults to `null`.

## dragSnap

Snapping unit when dragging items. Defaults to `15 * 60 * 1000` or 15min. When so, the items will snap to 15min intervals when dragging.

## minResizeWidth

The minimum width, in pixels, of a timeline entry when it's possible to resize. If not reached, you must zoom in to resize more. Default to `20`.

## lineHeight

Height of one line in the calendar in pixels. Default `30`

## itemHeightRatio

What percentage of the height of the line is taken by the item? Default `0.65`

## minZoom

Smallest time the calendar can zoom to in milliseconds. Default `60 * 60 * 1000` (1 hour)

**notes**: please note than second won't show up unless you change this to `60 * 1000`

## maxZoom

Largest time the calendar can zoom to in milliseconds. Default `5 * 365.24 * 86400 * 1000` (5 years)

## clickTolerance

How many pixels we can drag the background for it to be counted as a click on the background. Default `3`

## canMove

Can items be dragged around? Can be overridden in the `items` array. Defaults to `true`

## canChangeGroup

Can items be moved between groups? Can be overridden in the `items` array. Defaults to `true`

## canResize

Can items be resized? Can be overridden in the `items` array. Accepted values: `false`, `"left"`, `"right"`, `"both"`. Defaults to `"right"`. If you pass `true`, it will be treated as `"right"` to not break compatibility with versions 0.9 and below.

## useResizeHandle

Append a special `.rct-drag-right` handle to the elements and only resize if dragged from there. Defaults to `false`

### stackItems

Stack items under each other, so there is no visual overlap when times collide. Can be overridden in the `groups` array. Defaults to `false`. Requires millisecond or `Moment` timestamps, not native JavaScript `Date` objects.

## traditionalZoom

Zoom in when scrolling the mouse up/down. Defaults to `false`

## itemTouchSendsClick

Normally tapping (touching) an item selects it. If this is set to true, a tap will have the same effect, as selecting with the first click and then clicking again to open and send the onItemClick event. Defaults to `false`.

## timeSteps

With what step to display different units. E.g. `15` for `minute` means only minutes 0, 15, 30 and 45 will be shown.

Default:

```js
{
  second: 1,
  minute: 1,
  hour: 1,
  day: 1,
  month: 1,
  year: 1
}
```

## scrollRef

Ref callback that gets a DOM reference to the scroll body element. Can be useful to programmatically scroll.

## onItemDrag(itemDragObject)

Called when an item is moving or resizing. Returns an object with the following properties:

| property        | type     | description                                                           |
| --------------- | -------- | --------------------------------------------------------------------- |
| `eventType`     | `string` | retuns either `move` or `resize`                                      |
| `itemId`        | `number` | ID of the item being moved or resized                                 |
| `time`          | `number` | UNIX timestamp in milliseconds                                        |
| `edge`          | `string` | on `resize`, returns a value of either `left` or `right`              |
| `newGroupOrder` | `number` | on `move`, index position of the new group that the item is moving to |

## onItemMove(itemId, dragTime, newGroupOrder)

Callback when an item is moved. Returns 1) the item's ID, 2) the new start time and 3) the index of the new group in the `groups` array.

## onItemResize(itemId, time, edge)

Callback when an item is resized. Returns 1) the item's ID, 2) the new start or end time of the item 3) The edge that was dragged (`left` or `right`)

## onItemSelect(itemId, e, time)

Called when an item is selected. This is sent on the first click on an item. `time` is the time that corresponds to where you click/select on the item in the timeline.

## onItemDeselect(e)

Called when deselecting an item. Used to clear controlled selected prop.

## onItemClick(itemId, e, time)

Called when an item is clicked. Note: the item must be selected before it's clicked... except if it's a touch event and `itemTouchSendsClick` is enabled. `time` is the time that corresponds to where you click on the item in the timeline.

## onItemDoubleClick(itemId, e, time)

Called when an item was double clicked. `time` is the time that corresponds to where you double click on the item in the timeline.

## onItemContextMenu(itemId, e, time)

Called when the item is clicked by the right button of the mouse. `time` is the time that corresponds to where you context click on the item in the timeline. Note: If this property is set the default context menu doesn't appear.

## onCanvasClick(groupId, time, e)

Called when an empty spot on the canvas was clicked. Get the group ID and the time as arguments. For example open a "new item" window after this.

## onCanvasDoubleClick(groupId, time, e)

Called when an empty spot on the canvas was double clicked. Get the group ID and the time as arguments.

## onCanvasContextMenu(groupId, time, e)

Called when the canvas is clicked by the right button of the mouse. Note: If this property is set the default context menu doesn't appear

## onZoom(timelineContext, unit)

Called when the timeline is zoomed, either via mouse/pinch zoom or clicking header to change timeline units

## moveResizeValidator(action, itemId, time, resizeEdge)

This function is called when an item is being moved or resized. It's up to this function to return a new version of `change`, when the proposed move would violate business logic.

The argument `action` is one of `move` or `resize`.

The argument `resizeEdge` is when resizing one of `left` or `right`.

The argument `time` describes the proposed new time for either the start time of the item (for move) or the start or end time (for resize).

The function must return a new unix timestamp in milliseconds... or just `time` if the proposed new time doesn't interfere with business logic.

For example, to prevent moving of items into the past, but to keep them at 15min intervals, use this code:

```js

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres (more or less) to [Semantic Versioning](http://semver.org/).

## Unreleased

## 0.28.0

- Add unit argument to onZoom and onTimeChange callbacks @hckr #655
- Add `className` prop to Timeline component to override `react-calendar-timeline` class #682
- support zoom level seconds #835 @horizon-plaza
- custom buffer prop (help with controlled scrolling) @Ilaiwi
- Fix injecting custom vertical line's class names for time periods longer than day @RafikiTiki
- fix Context Menu unintentionally disabled by default @dsgipe #769
- delete props `headerLabelFormats` and `subHeaderLabelFormats` not you can pass `formatLabel` function to `DateHeader` with label width and start and end time of intervals

## 0.26.7

- fix scrolling with trackpad @ilaiwi #679
- remove duplicate proptype validation in `TimelineStateContext` @xat

## 0.26.6

- fix `visibleTimeStart`, `visibleTimeEnd` and `onTimeChange` not working as expected in controlled mode @ilaiwi

### examples

two new examples

#### Controlled scroll

Controlled visible port of the calendar using `visibleTimeStart` and `visibleTimeEnd`. This also limits scrolling by mouse and adds two buttons to change the visible port of the calendar

[Example Codesandbox](https://codesandbox.io/s/timeline-demo-controlled-visible-time-no-scroll-659jb)

#### Programmatically Scrolling

Using controlled scroll and react-spring to trigger scrolling and create an animation.

[Example Codesandbox](https://codesandbox.io/s/confident-waterfall-3kq2503y8p)

## 0.26.5

- improve performance by:
  - eliminate extra call of layout on state update @ilaiwi
  - eliminate unmounting and mounting of Interval Component @ilaiwi

## 0.26.4

- fix `react-calendar-timeline` not working with `react-hot-loader` #607 @ilaiwi + @westn
- add documentation for `stackItems` format #661 @tyson-kubota

## 0.26.3

- add documentation for `onItemDeselect` #350 @ilaiwi
- solve a bug where `onItemDeselect` is not triggered as expected for several item clicks #350 @ilaiwi
- fix row height on browser scaling #615 @gaston-niglia

### Packages

update to `node-sass@4.12.0` for newer versions of node.

## 0.26.2

- render the items layer after columns and rows for layring @ilaiwi

## 0.26.1

- fix issue where mouse down gets stuck when scrolling the timeline #526 @KhalidArdah

you can as well solve the issue without upgrading by adding the following style

```
.react-calendar-timeline .rct-horizontal-lines {
  -webkit-user-select: none;
  -moz-user-select: -moz-none;
  -ms-user-select: none;
  user-select: none;
}
```

[as here](https://codesandbox.io/s/timeline-demo-sticky-header-w6s5f)

## 0.26.0

#### Added

- Add `onItemDrag` prop to `<Timeline />` #517 @bettymakes
- Upgrade to Babel 7.5.0, Jest 24.8.0, Enzyme 3.10.0 @trevdor

#### Breaking

- Removed `<InfoLabel />` in favour of allowing for custom component to be rendered on move or resize. Check out the demo in `demo/app/demo-custom-info-label` for an example on how to display your own custom info label or [this example](https://codesandbox.io/s/timeline-demo-info-label-neec9).

## 0.25.4

- Move `classnames` to a production dependency

## 0.25.3

- Fixed the `undefined` classnames in TimelineHeaders #566 @trevdor

## 0.25.2

- Fixed the auto-scroll right bug in a scaled browser. #528 @cw196

## 0.25.1

- fix error when using `week` unit causing format error in `DateHeader` #562 @dkarnutsch
- fix Wheel/Mousewheel Event errors on chrome 73 #541 @ilaiwi

## 0.25.0

### Custom Headers

This new feature gives more control to dev to create customizable headers to provide better UI. Now user have more control through a set of new components to render headers. This new feature came with a breaking change though.

```jsx
import Timeline, { TimelineHeaders, SidebarHeader, DateHeader } from 'react-calendar-timeline';

<Timeline>
  <TimelineHeaders>
    <SidebarHeader>
      {({ getRootProps }) => {
        return <div {...getRootProps()}>Left</div>;
      }}
    </SidebarHeader>
    <DateHeader unit="primaryHeader" />
    <DateHeader />
    <CustomHeader height={50} headerData={{ someData: 'data' }} unit="year">
      {({ headerContext: { intervals }, getRootProps, getIntervalProps, showPeriod, data }) => {
        return (
          <div {...getRootProps()}>
            {intervals.map((interval) => {
              const intervalStyle = {
                lineHeight: '30px',
                textAlign: 'center',
                borderLeft: '1px solid black',
                cursor: 'pointer',
                backgroundColor: 'Turquoise',
                color: 'white',
              };
              return (
                <div
                  onClick={() => {
                    showPeriod(interval.startTime, interval.endTime);
                  }}
                  {...getIntervalProps({
                    interval,
                    style: intervalStyle,
                  })}
                >
                  <div className="sticky">{interval.startTime.format('YYYY')}</div>
                </div>
              );
            })}
          </div>
        );
      }}
    </CustomHeader>
  </TimelineHeaders>
</Timeline>;
```

Check out the new docs before please [here](https://github.com/namespace-ee/react-calendar-timeline/tree/custom-headers#timeline-headers)

#### removed props

- `stickyOffset` and `stickyHeader` now you can make your header sticky by following this [examples](https://github.com/namespace-ee/react-calendar-timeline/tree/master/examples#custom-item-rendering)
- `headerRef` to get the headerRef you need to pass ref callback to `TimelineHeader` component
- `headerLabelGroupHeight` and `headerLabelHeight` now you can pass a `height` prop to both `CustomHeader` and `DateHeader`

src/lib/items/Item.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import interact from 'interactjs';
import moment from 'moment';
import clsx from 'clsx';

import { _get, deepObjectCompare } from '../utility/generic';
import { composeEvents } from '../utility/events';
import { defaultItemRenderer } from './defaultItemRenderer';
import { coordinateToTimeRatio } from '../utility/calendar';
import { getSumScroll, getSumOffset } from '../utility/dom-helpers';
import {
  overridableStyles,
  selectedStyle,
  selectedAndCanMove,
  selectedAndCanResizeLeft,
  selectedAndCanResizeLeftAndDragLeft,
  selectedAndCanResizeRight,
  selectedAndCanResizeRightAndDragRight,
  leftResizeStyle,
  rightResizeStyle,
} from './styles';
import { TimelineStateConsumer } from '../timeline/TimelineStateContext';

export default class Item extends Component {
  static propTypes = {
    canvasTimeStart: PropTypes.number.isRequired,
    canvasTimeEnd: PropTypes.number.isRequired,
    canvasWidth: PropTypes.number.isRequired,
    order: PropTypes.object,
  
    dragSnap: PropTypes.number,
    minResizeWidth: PropTypes.number,
    selected: PropTypes.bool,
  
    canChangeGroup: PropTypes.bool.isRequired,
    canMove: PropTypes.bool.isRequired,
    canResizeLeft: PropTypes.bool.isRequired,
    canResizeRight: PropTypes.bool.isRequired,
  
    keys: PropTypes.object.isRequired,
    item: PropTypes.object.isRequired,
  
    onSelect: PropTypes.func,
    onDrag: PropTypes.func,
    onDrop: PropTypes.func,
    onResizing: PropTypes.func,
    onResized: PropTypes.func,
    onContextMenu: PropTypes.func,
    itemRenderer: PropTypes.func,
  
    itemProps: PropTypes.object,
    canSelect: PropTypes.bool,
    dimensions: PropTypes.object,
    groupTops: PropTypes.array,
    useResizeHandle: PropTypes.bool,
    moveResizeValidator: PropTypes.func,
    onItemDoubleClick: PropTypes.func,
  
    scrollRef: PropTypes.object,
  };

  // removed prop type check for SPEED!
  // they are coming from a trusted component anyway
  // (this complicates performance debugging otherwise)
  static defaultProps = {
    selected: false,
    itemRenderer: defaultItemRenderer,
  };

  constructor(props) {
    super(props);

    this.cacheDataFromProps(props);

    this.state = {
      interactMounted: false,

      dragging: null,
      dragStart: null,
      preDragPosition: null,
      dragTime: null,
      dragGroupDelta: null,

      resizing: null,
      resizeEdge: null,
      resizeStart: null,
      resizeTime: null,
    };
  }

  shouldComponentUpdate(nextProps, nextState) {
    const shouldUpdate =
      nextState.dragging !== this.state.dragging ||
      nextState.dragTime !== this.state.dragTime ||
      nextState.dragGroupDelta !== this.state.dragGroupDelta ||
      nextState.resizing !== this.state.resizing ||
      nextState.resizeTime !== this.state.resizeTime ||
      nextProps.keys !== this.props.keys ||
      !deepObjectCompare(nextProps.itemProps, this.props.itemProps) ||
      nextProps.selected !== this.props.selected ||
      nextProps.item !== this.props.item ||
      nextProps.canvasTimeStart !== this.props.canvasTimeStart ||
      nextProps.canvasTimeEnd !== this.props.canvasTimeEnd ||
      nextProps.canvasWidth !== this.props.canvasWidth ||
      (nextProps.order ? nextProps.order.index : undefined) !==
        (this.props.order ? this.props.order.index : undefined) ||
      nextProps.dragSnap !== this.props.dragSnap ||
      nextProps.minResizeWidth !== this.props.minResizeWidth ||
      nextProps.canChangeGroup !== this.props.canChangeGroup ||
      nextProps.canSelect !== this.props.canSelect ||
      nextProps.canMove !== this.props.canMove ||
      nextProps.canResizeLeft !== this.props.canResizeLeft ||
      nextProps.canResizeRight !== this.props.canResizeRight ||
      nextProps.dimensions !== this.props.dimensions;
    return shouldUpdate;
  }

  cacheDataFromProps(props) {
    this.itemId = _get(props.item, props.keys.itemIdKey);
    this.itemTitle = _get(props.item, props.keys.itemTitleKey);
    this.itemDivTitle = props.keys.itemDivTitleKey ? _get(props.item, props.keys.itemDivTitleKey) : this.itemTitle;
    this.itemTimeStart = _get(props.item, props.keys.itemTimeStartKey);
    this.itemTimeEnd = _get(props.item, props.keys.itemTimeEndKey);
  }

  getTimeRatio() {
    const { canvasTimeStart, canvasTimeEnd, canvasWidth } = this.props;
    return coordinateToTimeRatio(canvasTimeStart, canvasTimeEnd, canvasWidth);
  }

  dragTimeSnap(dragTime, considerOffset) {
    const { dragSnap } = this.props;
    if (dragSnap) {
      const offset = considerOffset ? moment().utcOffset() * 60 * 1000 : 0;
      return Math.round(dragTime / dragSnap) * dragSnap - (offset % dragSnap);
    } else {
      return dragTime;
    }
  }

  resizeTimeSnap(dragTime) {
    const { dragSnap } = this.props;
    if (dragSnap) {
      const endTime = this.itemTimeEnd % dragSnap;
      return Math.round((dragTime - endTime) / dragSnap) * dragSnap + endTime;
    } else {
      return dragTime;
    }
  }

  dragTime(e) {
    const startTime = moment(this.itemTimeStart);

    if (this.state.dragging) {
      return this.dragTimeSnap(this.timeFor(e) + this.state.dragStart.offset, true);
    } else {
      return startTime;
    }
  }

  timeFor(e) {
    const ratio = coordinateToTimeRatio(this.props.canvasTimeStart, this.props.canvasTimeEnd, this.props.canvasWidth);

    const offset = getSumOffset(this.props.scrollRef).offsetLeft;
    const scrolls = getSumScroll(this.props.scrollRef);

    return (e.pageX - offset + scrolls.scrollLeft) * ratio + this.props.canvasTimeStart;
  }

  dragGroupDelta(e) {
    const { groupTops, order } = this.props;
    if (this.state.dragging) {
      if (!this.props.canChangeGroup) {
        return 0;
      }
      let groupDelta = 0;

      const offset = getSumOffset(this.props.scrollRef).offsetTop;
      const scrolls = getSumScroll(this.props.scrollRef);

      for (let key of Object.keys(groupTops)) {
        const groupTop = groupTops[key];
        if (e.pageY - offset + scrolls.scrollTop > groupTop) {
          groupDelta = parseInt(key, 10) - order.index;
        } else {
          break;
        }
      }

      if (this.props.order.index + groupDelta < 0) {
        return 0 - this.props.order.index;
      } else {
        return groupDelta;
      }
    } else {
      return 0;
    }
  }

  resizeTimeDelta(e, resizeEdge) {
    const length = this.itemTimeEnd - this.itemTimeStart;
    const timeDelta = this.dragTimeSnap((e.pageX - this.state.resizeStart) * this.getTimeRatio());

    if (length + (resizeEdge === 'left' ? -timeDelta : timeDelta) < (this.props.dragSnap || 1000)) {
      if (resizeEdge === 'left') {
        return length - (this.props.dragSnap || 1000);
      } else {
        return (this.props.dragSnap || 1000) - length;
      }
    } else {
      return timeDelta;
    }
  }

  onDragStart(e) {
    if (this.props.selected) {
      const clickTime = this.timeFor(e);
      this.setState({
        dragging: true,
        dragStart: {
          x: e.pageX,
          y: e.pageY,
          offset: this.itemTimeStart - clickTime,
        },
        preDragPosition: { x: e.target.offsetLeft, y: e.target.offsetTop },
        dragTime: this.itemTimeStart,
        dragGroupDelta: 0,
      });
    } else {
      return false;
    }
  }

  onDragMove(e) {
    if (this.state.dragging) {
      let dragTime = this.dragTime(e);
      let dragGroupDelta = this.dragGroupDelta(e);
      if (this.props.moveResizeValidator) {
        dragTime = this.props.moveResizeValidator('move', this.props.item, dragTime);
      }

      if (this.props.onDrag) {
        this.props.onDrag(this.itemId, dragTime, this.props.order.index + dragGroupDelta);
      }

      this.setState({
        dragTime: dragTime,
        dragGroupDelta: dragGroupDelta,
      });
    }
  }

  onDragEnd(e) {
    if (this.state.dragging) {
      if (this.props.onDrop) {
        let dragTime = this.dragTime(e);

        if (this.props.moveResizeValidator) {
          dragTime = this.props.moveResizeValidator('move', this.props.item, dragTime);
        }

        this.props.onDrop(this.itemId, dragTime, this.props.order.index + this.dragGroupDelta(e));
      }

      this.setState({
        dragging: false,
        dragStart: null,
        preDragPosition: null,
        dragTime: null,
        dragGroupDelta: null,
      });
    }
  }

  onResizeStart(e) {
    if (this.props.selected) {
      this.setState({
        resizing: true,
        resizeEdge: null, // we don't know yet
        resizeStart: e.pageX,
        resizeTime: 0,
      });
    } else {
      return false;
    }
  }

  onResizeMove(e) {
    if (this.state.resizing) {
      let resizeEdge = this.state.resizeEdge;

      if (!resizeEdge) {
        resizeEdge = e.deltaRect.left !== 0 ? 'left' : 'right';
        this.setState({ resizeEdge });
      }
      let resizeTime = this.resizeTimeSnap(this.timeFor(e));

      if (this.props.moveResizeValidator) {
        resizeTime = this.props.moveResizeValidator('resize', this.props.item, resizeTime, resizeEdge);
      }

      if (this.props.onResizing) {
        this.props.onResizing(this.itemId, resizeTime, resizeEdge);
      }

      this.setState({
        resizeTime,
      });
    }
  }

  onResizeEnd(e) {
    if (this.state.resizing) {
      const { resizeEdge } = this.state;
      let resizeTime = this.resizeTimeSnap(this.timeFor(e));

      if (this.props.moveResizeValidator) {
        resizeTime = this.props.moveResizeValidator('resize', this.props.item, resizeTime, resizeEdge);
      }

      if (this.props.onResized) {
        this.props.onResized(this.itemId, resizeTime, resizeEdge, this.resizeTimeDelta(e, resizeEdge));
      }
      this.setState({
        resizing: null,
        resizeStart: null,
        resizeEdge: null,
        resizeTime: null,
      });
    }
  }

  mountInteract() {
    const leftResize = this.props.useResizeHandle ? '.rct-item-handler-resize-left' : true;
    const rightResize = this.props.useResizeHandle ? '.rct-item-handler-resize-right' : true;

    interact(this.item)
      .resizable({
        edges: {
          left: this.canResizeLeft() && leftResize,
          right: this.canResizeRight() && rightResize,
          top: false,
          bottom: false,
        },
        enabled: this.props.selected && (this.canResizeLeft() || this.canResizeRight()),
      })
      .draggable({
        enabled: this.props.selected && this.canMove(),
      })
      .styleCursor(false)
      .on('dragstart', this.onDragStart.bind(this))
      .on('dragmove', this.onDragMove.bind(this))
      .on('dragend', this.onDragEnd.bind(this))
      .on('resizestart', this.onResizeStart.bind(this))
      .on('resizemove', this.onResizeMove.bind(this))
      .on('resizeend', this.onResizeEnd.bind(this))
      .on('tap', (e) => {
        this.actualClick(e, e.pointerType === 'mouse' ? 'click' : 'touch');
      });

    this.setState({
      interactMounted: true,
    });
  }

  canResizeLeft(props = this.props) {
    if (!props.canResizeLeft) {
      return false;
    }
    let width = parseInt(props.dimensions.width, 10);
    return width >= props.minResizeWidth;
  }

  canResizeRight(props = this.props) {
    if (!props.canResizeRight) {
      return false;
    }
    let width = parseInt(props.dimensions.width, 10);
    return width >= props.minResizeWidth;
  }

  canMove(props = this.props) {
    return !!props.canMove;
  }

  componentDidUpdate(prevProps) {
    this.cacheDataFromProps(this.props);
    const couldDrag = prevProps.selected && this.canMove(prevProps);
    const couldResizeLeft = prevProps.selected && this.canResizeLeft(prevProps);
    const couldResizeRight = prevProps.selected && this.canResizeRight(prevProps);
    const willBeAbleToDrag = this.props.selected && this.canMove(this.props);
    const willBeAbleToResizeLeft = this.props.selected && this.canResizeLeft(this.props);
    const willBeAbleToResizeRight = this.props.selected && this.canResizeRight(this.props);

    if (!this.item) {
      this.setState({
        interactMounted: false,
      });
      return;
    }

    let { interactMounted } = this.state;

    if (this.props.selected && !interactMounted) {
      this.mountInteract();
      interactMounted = true;
    }

    if (
      interactMounted &&
      (couldResizeLeft !== willBeAbleToResizeLeft || couldResizeRight !== willBeAbleToResizeRight)
    ) {
      const leftResize = this.props.useResizeHandle ? this.dragLeft : true;
      const rightResize = this.props.useResizeHandle ? this.dragRight : true;

      interact(this.item).resizable({
        enabled: willBeAbleToResizeLeft || willBeAbleToResizeRight,
        edges: {
          top: false,
          bottom: false,
          left: willBeAbleToResizeLeft && leftResize,
          right: willBeAbleToResizeRight && rightResize,
        },
      });
    }
    if (interactMounted && couldDrag !== willBeAbleToDrag) {
      interact(this.item).draggable({ enabled: willBeAbleToDrag });
    }

    this.setState({
      interactMounted,
    });
  }

  onMouseDown = (e) => {
    if (!this.state.interactMounted) {
      e.preventDefault();
      this.startedClicking = true;
    }
  };

  onMouseUp = (e) => {
    if (!this.state.interactMounted && this.startedClicking) {
      this.startedClicking = false;
      this.actualClick(e, 'click');
    }
  };

  onTouchStart = (e) => {
    if (!this.state.interactMounted) {
      e.preventDefault();
      this.startedTouching = true;
    }
  };

  onTouchEnd = (e) => {
    if (!this.state.interactMounted && this.startedTouching) {
      this.startedTouching = false;
      this.actualClick(e, 'touch');
    }
  };

  handleDoubleClick = (e) => {
    e.stopPropagation();
    if (this.props.onItemDoubleClick) {
      this.props.onItemDoubleClick(this.itemId, e);
    }
  };

  handleContextMenu = (e) => {
    if (this.props.onContextMenu) {
      e.preventDefault();
      e.stopPropagation();
      this.props.onContextMenu(this.itemId, e);
    }
  };

  actualClick(e, clickType) {
    if (this.props.canSelect && this.props.onSelect) {
      this.props.onSelect(this.itemId, clickType, e);
    }
  }

  getItemRef = (el) => (this.item = el);
  getDragLeftRef = (el) => (this.dragLeft = el);
  getDragRightRef = (el) => (this.dragRight = el);

  getItemProps = (props = {}) => {
    //TODO: maybe shouldnt include all of these classes
    const classNames = 'rct-item' + (this.props.item.className ? ` ${this.props.item.className}` : '');

    return {
      key: this.itemId,
      ref: this.getItemRef,
      title: this.itemDivTitle,
      className: classNames + ` ${props.className ? props.className : ''}`,
      onMouseDown: composeEvents(this.onMouseDown, props.onMouseDown),
      onMouseUp: composeEvents(this.onMouseUp, props.onMouseUp),
      onTouchStart: composeEvents(this.onTouchStart, props.onTouchStart),
      onTouchEnd: composeEvents(this.onTouchEnd, props.onTouchEnd),
      onDoubleClick: composeEvents(this.handleDoubleClick, props.onDoubleClick),
      onContextMenu: composeEvents(this.handleContextMenu, props.onContextMenu),
      style: this.getItemStyle(props),
    };
  };

  getResizeProps = (props = {}) => {
    return {
      left: {
        ref: this.getDragLeftRef,
        className: clsx('rct-item-handler rct-item-handler-left rct-item-handler-resize-left', props.leftClassName),
        style: {
          ...leftResizeStyle,
          ...props.leftStyle,
        },
      },
      right: {
        ref: this.getDragRightRef,
        className: clsx('rct-item-handler rct-item-handler-right rct-item-handler-resize-right', props.rightClassName),
        style: {
          ...rightResizeStyle,
          ...props.rightStyle,
        },
      },
    };
  };

  getItemStyle(props) {
    const dimensions = this.props.dimensions;

    return {
      ...overridableStyles,
      ...(this.props.selected && { ...selectedStyle }),
      ...(this.props.selected && this.canMove(this.props) && { ...selectedAndCanMove }),
      ...(this.props.selected && this.canResizeLeft(this.props) && { ...selectedAndCanResizeLeft }),
      ...(this.props.selected &&
        this.canResizeLeft(this.props) &&
        this.state.dragging && { ...selectedAndCanResizeLeftAndDragLeft }),
      ...(this.props.selected & this.canResizeRight(this.props) && { ...selectedAndCanResizeRight }),
      ...(this.props.selected & this.canResizeRight(this.props) & this.state.dragging && {
        ...selectedAndCanResizeRightAndDragRight,
      }),
      ...props.style,
      position: 'absolute',
      left: `${dimensions.left}px`,
      top: `${dimensions.top}px`,
      width: `${dimensions.width}px`,
      height: `${dimensions.height}px`,
      lineHeight: `${dimensions.height}px`,
    };
  }

  render() {
    if (typeof this.props.order === 'undefined' || this.props.order === null) {
      return null;
    }

    const itemContext = {
      dimensions: this.props.dimensions,
      useResizeHandle: this.props.useResizeHandle,
      title: this.itemTitle,
      canMove: this.canMove(this.props),
      canResizeLeft: this.canResizeLeft(this.props),
      canResizeRight: this.canResizeRight(this.props),
      selected: this.props.selected,
      dragging: this.state.dragging,
      dragStart: this.state.dragStart,
      dragTime: this.state.dragTime,
      dragGroupDelta: this.state.dragGroupDelta,
      resizing: this.state.resizing,
      resizeEdge: this.state.resizeEdge,
      resizeStart: this.state.resizeStart,
      resizeTime: this.state.resizeTime,
      width: this.props.dimensions.width,
    };

    return (
      <TimelineStateConsumer>
        {({ getTimelineState }) => {
          return this.props.itemRenderer({
            item: this.props.item,
            timelineContext: getTimelineState(),
            itemContext,
            getItemProps: this.getItemProps,
            getResizeProps: this.getResizeProps,
          });
        }}
      </TimelineStateConsumer>
    );
  }
}
